### DISCUSSION TITLE
Best practices for clean-up after segfault / custom signal handler

### DISCUSSION THREAD

**Discussion by daniel-falk** [getsentry/sentry-native#848] [Q&A] [answered]
We have some peripheral clean-up that we need to do before the application is closed down. This should be done even if there is e.g. a SEGFAULT. We do this currently by adding a signal handler for SEGFAULT that sets an error flag, we then do clean-up and exits the application.

When using sentry-native it seems like the library is overriding our signal handlers to be able to send telemetry messages to sentry on SEGFAULTs. This handler seems to exit the application when done, meaning that we can't do our cleanup. We could manually override the sentry signal handler, but then we don't get the telemetry messages. Do we need to manually implement the transmission of a stacktrace to using sentry from our own signal handler, or is there a way we can get sentry to remove it's handler and raise the same (or another) signal after sending the stacktrace? Or is there any better way to design the error handling to achieve both clean-up and telemetry?

**Comment by supervacuus** [comment, marked as answer]
Hi, @daniel-falk! If you are asking for a solution specifically when a crash occurs, the preferred method would be to implement your shutdown by using our `on_crash`-hook which will be called from the crash handler before the crash snapshot is generated (not only for `SIGSEGV` but all hard-exceptions). 

Here is the link to the docs: https://docs.sentry.io/platforms/native/configuration/filtering/#using-on_crash
And a bit more detailed in the header docs: https://github.com/getsentry/sentry-native/blob/6c76f5845491e6b79711a34715523fc2b800f614/include/sentry.h#L831-L875

**Reply by tian-yu-ocado** [reply]
Hi @supervacuus, I have a follow up question with respect to use `sentry_options_set_on_crash`. I have a cleanup callback inside which I would like to change the title of the event, as in my Sentry web portal the reported event title is <unknown>. Is there a way to do this inside the `cleanup` callback function? What I did was like this:
```
static sentry_value_t crash_cleanup(
        const sentry_ucontext_t* uctx,  // provides the user-space context of the crash
        sentry_value_t event,           // used the same way as in `before_send`
        void* closure                   // user-data that you can provide at configuration time
) {
    sentry_value_set_by_key(event, "title", sentry_value_new_string("crash-title"));
    sentry_value_set_by_key(event, "message", sentry_value_new_string("stack crashed"));       
    return event;
}
```
However, setting the message seem to work properly, but setting the event title does not.

**Reply by supervacuus** [reply]
AFAIK, there is no way to change the "title" from the client because it is a computed property of the backend. "message" OTOH is part of the [Message Interface](https://develop.sentry.dev/sdk/event-payloads/message/) and can be sent from any client.

**Reply by tian-yu-ocado** [reply]
Thanks! If the event title is computed by the backend, do you happen to know what information is used to do the calculation? For my case if I trigger a hard crash by nullptr dereferncing, the event title would just be <unknown>. Would it because it is missing the debug information?

**Reply by supervacuus** [reply]
The "title" typically reflects the function/symbol of the top-level frame in the exception stack trace. So, if that symbol cannot be resolved, it might be empty (but I don't know the details of the current implementation).

**Reply by tian-yu-ocado** [reply]
Thanks a lot! @supervacuus
